### ISSUE TITLE
Shape of gradient unfit for keras?

### ISSUE THREAD

**Issue by githubnemo** [mcf06/theano_ctc#10]
I noticed that the gradient from warp-ctc is not touched by the wrapper in any way so it has the same dimensions as the input activations, i.e. `(seqlen, batches, units)`.

Doesn't this conflict with every other gradient shape used in keras?

Test file: https://github.com/githubnemo/theano_ctc/blob/gradients/tests/test_ctc_gradients.py

**Comment by mcf06** [comment]
The gradient G of a scalar c with respect to a tensor X should be a tensor of partial derivatives with the same shape as x: G[a,b,c] = delta c / delta X[a,b,c]. Thus, I believe the CTC Ops do the right thing.

I don't use Keras, so I can't provide any comment there. In addition to warp-ctc, Theano's scan does require the sequence dimension to be the first dimension, so these are the activation dimensions that I'm familiar with.

**Comment by githubnemo** [comment]
It is doing the right thing on its own but I have doubts about compatibility with systems that use shape `(batches, seqlen, units)`. If I use a wrapper to use CTC in keras (as seen in [tests/test_keras.py](https://github.com/mcf06/theano_ctc/blob/master/tests/test_keras.py#L26))  that does a reshape of the activations to match the format of warp-ctc the computed CTC gradient will be the only gradient with shape `(seqlen, batches, units)`. 

This would not be a problem if I could isolate the gradient and then do a `dimshuffle` but since the gradient computation is not accessible in Keras (why would it be?) this is out of the question.

I tried to internalize the `dimshuffle` to test if this has any effect on the outcome but haven't succeeded (again, haven't understood the internals of Theano yet).

**Comment by mcf06** [comment]
Ops are contractually required to both forward and backpropagate their changes. So DimShuffle will shuffle its input to produce its output, but also reverse-shuffle its output_grads to produce grads wrt each input.

You can try it yourself in tests/test_ctc2.py. Change:

tsWeights = theano.shared(weights, name="weights")
tActs = T.dot(tsInputs, tsWeights)

to:

weights = weights.transpose()
tsWeights = theano.shared(weights, name="weights")
tActs = T.dot(tsInputs, tsWeights.dimshuffle((1,0)))

Now weights and tsWeights are 5x6 instead of 6x5, but so is the gradient wrt tsWeights. 

Under the hood, I expect K.permute_dimensions is doing the same thing.

**Comment by githubnemo** [comment]
Yep, I did this just a few hours ago and came to the same conclusion. Thank you for your input.

 I _think_ this can be closed as the real issue I have seems to be caused by #11 and the actual gradient computation is fine.
